controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package twophase
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/utils"
+)
+
+const ITER_MAX = 1e5
+
+type chaosStateMachine struct {
+	Chaos v1alpha1.InnerSchedulerObject
+	Req   ctrl.Request
+	*Reconciler
+}
+
+func unexpectedInto(currentPhase v1alpha1.ExperimentPhase, targetPhase v1alpha1.ExperimentPhase) error {
+	return errors.Errorf("turn from %s into %s is unexpected", currentPhase, targetPhase)
+}
+
+// This method changes the phase of an object and do some side effects
+// There are 6 different phases, so there could be 6 * 6 = 36 branches
```

Comment:
For which state could trans to the target state or not, this method seems still not clear to me.

Look at some well-designed fsm repo, e.g. [this one](https://github.com/looplab/fsm), it will setup fsm rules **explicitly**. In our scenario, `currentPhase` is the fsm state, and `targetPhase` is fsm target state, how about write down these 6*6 branches as a rule map, and use `canTrans` such grammar to check if the currentState could trans to the target state and do the job.

If we choose this implementation, we could remove these if-else statements, and it should be more clearer to our developers.

Reply:
`map` is a good idea 👍